src/main.rs

```diff
@@ -1,28 +1,177 @@
+                        )
+                    ) 
+                    .collect::<Result<Vec<_>>>()?,
+                None => bail!("no server addreses"),
+            },
+        },
+        _ => bail!("unknown subcommand"),
+    };
+
+    let local_ip = "127.0.0.1"
+        .parse::<IpAddr>()
+        .chain_err(|| "error parsing 127.0.0.1")?;
+
+    let local_addr = SocketAddr::new(
+        local_ip, 
+        match server {
+            Server::Forward{port} => port,
+            Server::Reverse{port, ..} => port,
+        }
+    );
+
+    let listen_socket = TcpListener::bind(&local_addr)
+        .chain_err(|| "error binding to local address")?;
+
+    match server {
+        Server::Forward{..} => forward_proxy::run(
+                listen_socket, 
+                compress, 
+                encrypt
+            )
+            .chain_err(|| "error in forward_proxy::run()")?,
+        Server::Reverse{server_ips, ..} => reverse_proxy::run(
+                listen_socket, 
+                server_ips, 
+                compress, 
+                encrypt
+            )
+            .chain_err(|| "error in reverse_proxy::run()")?,
+    }
+
+    return Ok(())
+}
 
 fn main() {
-    // General flow:
-    //  - parse command line arguments:
-    //      - run as forward or reverse proxy depending on args
-    //      - optionally compress data before encrypting (vulnerable to CRIME-style attacks)
-    //  - accept new TCP connections, handle each one in a new thread
-    //
-    //  - If forward proxy (runs for each new connection):
-    //      - Check to see if payload already includes TLS records, if so, forward the packet without modification
-    //      - Create a new TLS client session (assuming the use of rustls)
-    //      - Create a new TCP connection to the target server
-    //      - Loop until either client or server connection closes:
-    //          - If data received from client, write to server using write() and then write_tls() (assuming rustls)
-    //              - Optionally compress data before sending
-    //          - If data received from server, read the data using read_tls() and then read(), then send to client (assuming rustls)
-    //
-    //  - If reverse proxy (runs for each new connection):
-    //      - Check to see if payload includes TLS records, if not, forward the packet without modification
-    //      - Create a new TLS server session (assuming rustls)
-    //      - Create a new TCP connection to the target server
-    //      - Loop until either client or server connection closes:
-    //          - If data received from client, read the data using read_tls() and then read(), then send to server (assuming rustls)
-    //          - If data received from server, send to client using write() and then write_tls() (assuming rustls)
-    //              - Optionally compress data before sending
+    if let Err(e) = run() {
+        eprintln!("{}", e);
+
+        for e in e.iter().skip(1) {
```

Comment:
instead of printing the backtrace yourself, I believe you can use `e.display_chain().to_string()` to get the nicely formatted backtrace and error message